### PR TITLE
Implementation of get_key and get_all_key from :application in Elixir

### PR DESCRIPTION
**Pull request by romul**
According to discussion at https://github.com/phoenixframework/phoenix/pull/1147

### CODE REVIEW

File: lib/elixir/lib/application.ex [hunk 1]
```diff
@@ -193,6 +193,31 @@ defmodule Application do
...
+    end
+  end
+
+  @doc """
+  Returns the value for `key` in `app`'s specification.
+
+  If the configuration parameter does not exist, the function returns :undefined.
+  """
+  @spec get_key(app, key) :: value | :error
+  def get_key(app, key) do
```
Comment (josevalim) [inline lib/elixir/lib/application.ex]: We should pass the default value as argument or rename it to `fetch_key` and return `{:ok, value} | :error`.

Comment (josevalim) [inline lib/elixir/lib/application.ex]: Actually, we should probably have both get_key and fetch_key. get_key can be implemented on top of  fetch_key.

**Comment by josevalim** [comment]
Thank you, i have added some comments!

**Comment by romul** [comment]
Fixed.

File: lib/elixir/lib/application.ex [hunk 2]
```diff
@@ -193,6 +193,45 @@ defmodule Application do
...
+    case :application.get_all_key(app) do
+      {:ok, dict} -> dict
+      :undefined -> []
+    end
+  end
+
+  @doc """
+  Returns the value for `key` in `app`'s specification.
+
+  If the configuration parameter does not exist, the function returns the
```
Comment (lexmag) [inline lib/elixir/lib/application.ex]: I think it should say __specification__ parameter.

File: lib/elixir/lib/application.ex [hunk 3]
```diff
@@ -193,6 +193,45 @@ defmodule Application do
...
+    case fetch_key(app, key) do
+      {:ok, value} -> value
+      :error -> default
+    end
+  end
+
+  @doc """
+  Returns the value for `key` in `app`'s specification in a tuple.
+
+  If the configuration parameter does not exist, the function returns `:error`.
```
Comment (lexmag) [inline lib/elixir/lib/application.ex]: And same here.

**Comment by josevalim** [comment]
Thank you! @fishcakez please review too.

**Comment by josevalim** [comment]
Folks, I am wondering if `get_key` is a good name. It is very easy to mistake `get_env` by `get_key` and the error will be subtle. I am thinking we should try to call it something else. Maybe `Application.spec` and have a single function that always return all. Thoughts?

Thanks @romul. This pull request is great as is! If we are going to change the name, don't  worry about it, I will merge this and rename later.

**Comment by fishcakez** [comment]
I think the PR is perfect as is. Perhaps we could be an explanation of the difference between `key` and `env`?

**Comment by josevalim** [comment]
The explanation is not enough, I am afraid, because I am simply going to write some code that accidentally uses `get_key` and it will be hard to notice until I realize I have the wrong function.

**Comment by fishcakez** [comment]
Ah ok. I don't think we can call the function `spec` because the `env` is up to date and not the `env` from the spec.

**Comment by josevalim** [comment]
@fishcakez oh really? the more you know™.

**Comment by josevalim** [comment]
`Application.metadata/0`?

**Comment by fishcakez** [comment]
> Application.metadata/0?

Sure

**Comment by josevalim** [comment]
One idea is to mimic the API behind `Code.get_docs/2`:

    @spec metadata(app, :all) :: Keyword.t
    @spec metadata(app, :vsn) :: char_list()
    @spec metadata(app, :description) :: char_list()
    ...

Thoughts?

**Comment by fishcakez** [comment]
Can `:all` clash? I think it would better to provide a way to get a single value to skip an ets:select getting all the `env`

**Comment by josevalim** [comment]
The amount of ETS lookups is the same:

https://github.com/erlang/otp/blob/maint/lib/kernel/src/application_controller.erl#L356
https://github.com/erlang/otp/blob/maint/lib/kernel/src/application_controller.erl#L396

All doesn't seem to conflict unless a field named `:all` is added.

**Comment by fishcakez** [comment]
There's an extra call here:
https://github.com/erlang/otp/blob/d3fd0f9eb313501e91fb816ddfd8e17cd9b07c70/lib/kernel/src/application_controller.erl#L407
https://github.com/erlang/otp/blob/d3fd0f9eb313501e91fb816ddfd8e17cd9b07c70/lib/kernel/src/application_controller.erl#L342-L344

Ah so it uses a predefined/undocumented list and ignores extra ones :(.

**Comment by josevalim** [comment]
This has been merged manually, thank you!

**Pull request closed without merging**